edward/inferences/variational_inference.py

```diff
@@ -56,22 +56,43 @@ def initialize(self, optimizer=None, var_list=None, use_prettytensor=False,
     if var_list is None:
       # Traverse random variable graphs to get default list of variables.
       var_list = set()
+      latent_var_list = set()
+      data_var_list = set()
       trainables = tf.trainable_variables()
       for z, qz in six.iteritems(self.latent_vars):
         if isinstance(z, RandomVariable):
           var_list.update(get_variables(z, collection=trainables))
+          latent_var_list.update(get_variables(z, collection=trainables))
 
         var_list.update(get_variables(qz, collection=trainables))
 
       for x, qx in six.iteritems(self.data):
         if isinstance(x, RandomVariable) and \
                 not isinstance(qx, RandomVariable):
           var_list.update(get_variables(x, collection=trainables))
+          data_var_list.update(get_variables(x, collection=trainables))
 
       var_list = list(var_list)
 
     self.loss, grads_and_vars = self.build_loss_and_gradients(var_list)
 
+    if self.logging:
+      #TODO: when var_list is not None
+      tf.summary.scalar("loss", self.loss)
+      for var in latent_var_list:
+        for grad_var_tuple in grads_and_vars:
+          if var == grad_var_tuple[1]:
+            tf.summary.histogram("InferenceWeight_"+var.name, var)
+            tf.summary.histogram("InferenceGradient_" + var.name, grad_var_tuple[0])
+
+      for var in data_var_list:
+        for grad_var_tuple in grads_and_vars:
+          if var == grad_var_tuple[1]:
+            tf.summary.histogram("ModelWeight_" + var.name, var)
+            tf.summary.histogram("ModelGradient_" + var.name, grad_var_tuple[0])
+
+      self.summarize = tf.summary.merge_all()
```

Comment:
`tf.merge` requires storing all the summary objects. However, we build summaries in various methods of the Inference class, so this is not very TensorFlow-y. 

We can use a unique graph collection for each inference; then we can merge all summaries within that collection. For example:
```python
# summaries are added like this
summary_collection = 'summaries_' + str(id(inference))
tf.summary.scalar(..., collections=[summary_collection])
tf.summary.scalar(..., collections=[summary_collection])
# when forming merge summary
summary_collection = 'summaries_' + str(id(inference))
self.summarize = tf.summary.merge_all(summary_collection)
```